### ISSUE TITLE
Merge datastore and libp2p-peerstore

### ISSUE THREAD

**Issue by tomaka** [libp2p/rust-libp2p#328]
There's no reason to have `datastore` in its own crate.

**Comment by twittner** [comment]
While at it, you would not be interested in [twittner:datastore-queries](https://github.com/libp2p/rust-libp2p/compare/master...twittner:datastore-queries), wouldn't you?

**Comment by adamski** [comment]
I'm pretty much lurking here at the moment, but would if these were merged would it affect the ability to plug in different underlying data stores?

**Comment by tomaka** [comment]
Right now there's a JSON datastore, a memory datastore, a JSON peerstore, and a memory peerstore.
Removing the datastores doesn't remove the possibility to choose the way you want to store your peers.

The problem of the `datastore` crate is that its design is way too generic, and therefore inefficient. The memory datastore is nothing more than a hashmap, and the JSON datastore simply loads everything into a hashmap because there's no way it can do an efficient lookup in a file when we need a specific entry.

**Comment by adamski** [comment]
Yes, what I'm looking for eventually is being able to have a datastore that can be efficiently queried locally, basically a local KV/JSON database.

**Comment by tomaka** [comment]
No longer relevant after #723
